File: apps/web/src/lib/protocol.ts
import Ajv from 'ajv';

export type RoomPhase =
  | 'lobby'
  | 'collecting_entropy'
  | 'playing'
  | 'settlement'
  | 'paused'
  | 'closed';

export type ConnectionStatus = 'idle' | 'connecting' | 'connected' | 'reconnecting' | 'offline';
export type PlayerStatus = 'waiting' | 'active' | 'folded' | 'all_in' | 'sitting_out' | 'eliminated';
export type CardCode = `${string}${'c' | 'd' | 'h' | 's'}`;

export interface RoomConfig {
  maxPlayers: number;
  smallBlind: number;
  bigBlind: number;
  initialStack: number;
  actionTimeoutSeconds: 15 | 30 | 60;
}

export interface RoomSession {
  roomId: string;
  roomCode: string;
  memberId: string;
  nickname: string;
}

export interface PlayerState {
  memberId: string;
  nickname: string;
  seat: number;
  stack: number;
  borrowedTotal: number;
  rankingScore: number;
  streetBet: number;
  committed: number;
  status: PlayerStatus;
  ready: boolean;
  online: boolean;
  isHost: boolean;
  rebuyPending: boolean;
  holeCards?: CardCode[];
  lastAction?: string;
}

export interface LegalActions {
  canFold: boolean;
  canCheck: boolean;
  canCall: boolean;
  canRaise: boolean;
  canAllIn: boolean;
  callAmount: number;
  minRaiseTo: number;
  maxRaiseTo: number;
}

export interface PotState {
  amount: number;
  eligibleMemberIds: string[];
}

export interface WinnerState {
  memberId: string;
  nickname: string;
  amount: number;
  handName?: string;
}

export interface SettlementState {
  winners: WinnerState[];
  reason?: string;
}

export interface CompletedHandPlayer {
  memberId: string;
  nickname: string;
  seat: number;
  holeCards: CardCode[];
  handName: string;
  delta: number;
  folded: boolean;
}

export interface CompletedHand {
  handId: string;
  handNumber: number;
  completedAt: string;
  board: CardCode[];
  pot: number;
  players: CompletedHandPlayer[];
}

export interface RoomLogEntry {
  id: string;
  type: string;
  message: string;
  createdAt: string;
  handId?: string;
}

export interface FairnessState {
  serverCommitment?: string;
  merkleRoot?: string;
  contributionsReceived: number;
  contributionsRequired: number;
  contributionRequired: boolean;
  auditAvailable: boolean;
}

export interface GameSnapshot {
  roomId: string;
  roomCode: string;
  version: number;
  handId: string | null;
  turnId: string | null;
  phase: RoomPhase;
  config: RoomConfig;
  players: PlayerState[];
  buttonSeat: number | null;
  actingSeat: number | null;
  communityCards: CardCode[];
  holeCards: CardCode[];
  pots: PotState[];
  currentBet: number;
  legalActions: LegalActions;
  deadlineAt: string | null;
  serverNow: string;
  heroMemberId: string;
  lateJoin: boolean;
  paused: boolean;
  closePending: boolean;
  settlement?: SettlementState;
  roomLogs: RoomLogEntry[];
  completedHands: CompletedHand[];
  fairness: FairnessState;
}

export type RoomCommandType =
  | 'set_ready'
  | 'update_config'
  | 'start_hand'
  | 'contribute_randomness'
  | 'player_action'
  | 'pause_room'
  | 'resume_room'
  | 'remove_player'
  | 'close_room'
  | 'request_rebuy'
  | 'request_snapshot';

export interface RoomCommand<TPayload extends Record<string, unknown> = Record<string, unknown>> {
  commandId: string;
  roomId: string;
  handId: string | null;
  turnId: string | null;
  expectedVersion: number;
  type: RoomCommandType;
  payload: TPayload;
}

export interface CommandAck {
  status: 'accepted' | 'rejected';
  commandId: string;
  appliedVersion: number;
  errorCode?: string;
  message?: string;
  snapshot?: GameSnapshot;
}

export interface RoomEvent {
  version: number;
  type: string;
  snapshot?: GameSnapshot;
}

export interface SocketTicket {
  ticket: string;
  expiresAt?: string;
}

const ajv = new Ajv({ allErrors: true, allowUnionTypes: true });
const snapshotBoundarySchema = {
  type: 'object',
  required: ['roomId', 'roomCode', 'version', 'phase', 'players'],
  properties: {
    roomId: { type: 'string', minLength: 1 },
    roomCode: { type: 'string', minLength: 1 },
    version: { type: 'integer', minimum: 0 },
    phase: { type: 'string' },
    players: { type: 'array' },
  },
  additionalProperties: true,
} as const;
const validateSnapshotBoundary = ajv.compile(snapshotBoundarySchema);

const record = (value: unknown): Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value) ? value as Record<string, unknown> : {};

const first = (source: Record<string, unknown>, keys: string[]): unknown => {
  for (const key of keys) {
    if (source[key] !== undefined && source[key] !== null) return source[key];
  }
  return undefined;
};

const textValue = (source: Record<string, unknown>, keys: string[], fallback = ''): string => {
  const value = first(source, keys);
  return typeof value === 'string' ? value : fallback;
};

const numberValue = (source: Record<string, unknown>, keys: string[], fallback = 0): number => {
  const value = first(source, keys);
  return typeof value === 'number' && Number.isFinite(value) ? Math.trunc(value) : fallback;
};

const boolValue = (source: Record<string, unknown>, keys: string[], fallback = false): boolean => {
  const value = first(source, keys);
  return typeof value === 'boolean' ? value : fallback;
};

const listValue = (source: Record<string, unknown>, keys: string[]): unknown[] => {
  const value = first(source, keys);
  return Array.isArray(value) ? value : [];
};

const phaseAliases: Record<string, RoomPhase> = {
  lobby: 'lobby',
  waiting: 'lobby',
  collecting_entropy: 'collecting_entropy',
  collecting_randomness: 'collecting_entropy',
  shuffling: 'collecting_entropy',
  playing: 'playing',
  in_hand: 'playing',
  settlement: 'settlement',
  showdown: 'settlement',
  paused: 'paused',
  closed: 'closed',
};

const playerStatusAliases: Record<string, PlayerStatus> = {
  waiting: 'waiting',
  active: 'active',
  folded: 'folded',
  all_in: 'all_in',
  allin: 'all_in',
  sitting_out: 'sitting_out',
  sit_out: 'sitting_out',
  eliminated: 'eliminated',
};

const normalizeConfig = (value: unknown): RoomConfig => {
  const source = record(value);
  const timeout = numberValue(source, ['actionTimeoutSeconds', 'action_timeout_seconds', 'actionTimeout'], 30);
  return {
    maxPlayers: numberValue(source, ['maxPlayers', 'max_players'], 8),
    smallBlind: numberValue(source, ['smallBlind', 'small_blind'], 10),
    bigBlind: numberValue(source, ['bigBlind', 'big_blind'], 20),
    initialStack: numberValue(source, ['initialStack', 'initial_stack', 'buyIn'], 2000),
    actionTimeoutSeconds: timeout === 15 || timeout === 60 ? timeout : 30,
  };
};

const normalizePlayer = (value: unknown): PlayerState => {
  const source = record(value);
  const rawStatus = textValue(source, ['status'], 'waiting').toLowerCase();
  return {
    memberId: textValue(source, ['memberId', 'member_id', 'playerId', 'id']),
    nickname: textValue(source, ['nickname', 'name'], '玩家'),
    seat: numberValue(source, ['seat', 'seatIndex', 'seat_index']),
    stack: numberValue(source, ['stack', 'chips']),
    borrowedTotal: numberValue(source, ['borrowedTotal', 'borrowed_total']),
    rankingScore: numberValue(
      source,
      ['rankingScore', 'ranking_score'],
      numberValue(source, ['stack', 'chips']) - numberValue(source, ['borrowedTotal', 'borrowed_total']),
    ),
    streetBet: numberValue(source, ['streetBet', 'street_bet', 'bet']),
    committed: numberValue(source, ['committed', 'totalCommitted', 'total_committed']),
    status: playerStatusAliases[rawStatus] ?? 'waiting',
    ready: boolValue(source, ['ready', 'isReady', 'is_ready']),
    online: boolValue(source, ['online', 'connected', 'isOnline', 'is_online'], true),
    isHost: boolValue(source, ['isHost', 'is_host', 'host']),
    rebuyPending: boolValue(source, ['rebuyPending', 'rebuy_pending']),
    holeCards: normalizeCards(first(source, ['holeCards', 'hole_cards'])),
    lastAction: textValue(source, ['lastAction', 'last_action']) || undefined,
  };
};

const normalizeLegalActions = (value: unknown): LegalActions => {
  const source = record(value);
  const allowed = new Set(listValue(source, ['allowed', 'actions']).filter((item): item is string => typeof item === 'string'));
  return {
    canFold: boolValue(source, ['canFold', 'can_fold'], allowed.has('fold')),
    canCheck: boolValue(source, ['canCheck', 'can_check'], allowed.has('check')),
    canCall: boolValue(source, ['canCall', 'can_call'], allowed.has('call')),
    canRaise: boolValue(source, ['canRaise', 'can_raise'], allowed.has('raise_to') || allowed.has('raise')),
    canAllIn: boolValue(source, ['canAllIn', 'can_all_in'], allowed.has('all_in')),
    callAmount: numberValue(source, ['callAmount', 'call_amount']),
    minRaiseTo: numberValue(source, ['minRaiseTo', 'min_raise_to']),
    maxRaiseTo: numberValue(source, ['maxRaiseTo', 'max_raise_to']),
  };
};

const normalizeCards = (value: unknown): CardCode[] =>
  (Array.isArray(value) ? value : [])
    .map((card) => typeof card === 'string' ? card : textValue(record(card), ['code']))
    .filter((card): card is string => /^[2-9TJQKA][cdhs]$/i.test(card))
    .map((card) => `${card[0]?.toUpperCase()}${card[1]?.toLowerCase()}` as CardCode);

const normalizeCompletedHands = (value: unknown): CompletedHand[] => listValue(record(value), ['completedHands', 'completed_hands'])
  .map((entry) => {
    const source = record(entry);
    return {
      handId: textValue(source, ['handId', 'hand_id']),
      handNumber: numberValue(source, ['handNumber', 'hand_number']),
      completedAt: textValue(source, ['completedAt', 'completed_at']),
      board: normalizeCards(first(source, ['board'])),
      pot: numberValue(source, ['pot']),
      players: listValue(source, ['players']).map((item) => {
        const player = record(item);
        return {
          memberId: textValue(player, ['memberId', 'member_id']),
          nickname: textValue(player, ['nickname', 'name'], '玩家'),
          seat: numberValue(player, ['seat']),
          holeCards: normalizeCards(first(player, ['holeCards', 'hole_cards'])),
          handName: textValue(player, ['handName', 'hand_name'], '未摊牌'),
          delta: numberValue(player, ['delta']),
          folded: boolValue(player, ['folded']),
        };
      }),
    };
  })
  .filter((hand) => Boolean(hand.handId) && hand.handNumber > 0);

const normalizeRoomLogs = (value: unknown): RoomLogEntry[] => listValue(record(value), ['roomLogs', 'room_logs'])
  .map((entry) => {
    const source = record(entry);
    return {
      id: textValue(source, ['id']),
      type: textValue(source, ['type'], 'system'),
      message: textValue(source, ['message']),
      createdAt: textValue(source, ['createdAt', 'created_at']),
      handId: textValue(source, ['handId', 'hand_id']) || undefined,
    };
  })
  .filter((entry) => Boolean(entry.id) && Boolean(entry.message));

export const normalizeSnapshot = (value: unknown): GameSnapshot => {
  const outer = record(value);
  const source = record(first(outer, ['snapshot', 'data']) ?? outer);
  const handSource = record(first(source, ['hand']));
  const youSource = record(first(source, ['you']));
  const normalizedBoundary = {
    roomId: textValue(source, ['roomId', 'room_id']),
    roomCode: textValue(source, ['roomCode', 'room_code', 'code']).toUpperCase(),
    version: numberValue(source, ['version']),
    phase: textValue(source, ['phase', 'state'], 'lobby').toLowerCase(),
    players: listValue(source, ['players', 'members']),
  };
  if (!validateSnapshotBoundary(normalizedBoundary)) {
    throw new Error(`无效的房间快照：${ajv.errorsText(validateSnapshotBoundary.errors)}`);
  }

  const fairnessSource = record(first(source, ['fairness', 'shuffle']));
  const settlementSource = record(first(handSource, ['settlement']) ?? first(source, ['settlement', 'result']));
  const winnerValues = listValue(settlementSource, ['awards', 'winners']);
  const deadline = textValue(handSource, ['deadlineAt', 'deadline_at'], textValue(source, ['deadlineAt', 'deadline_at']));
  const phase = phaseAliases[normalizedBoundary.phase] ?? 'lobby';
  const normalizedPlayers = normalizedBoundary.players.map(normalizePlayer);
  const heroMemberId = textValue(youSource, ['memberId', 'member_id'], textValue(source, ['heroMemberId', 'hero_member_id', 'memberId', 'member_id']));
  const actorMemberId = textValue(handSource, ['actorMemberId', 'actor_member_id']);
  const heroPlayerSource = record(normalizedBoundary.players.find((value) => {
    const player = record(value);
    return textValue(player, ['memberId', 'member_id', 'playerId', 'id']) === heroMemberId;
  }));
  const explicitActorSeat = numberValue(handSource, ['actorSeat', 'actor_seat'], -1);
  const derivedActorSeat = normalizedPlayers.find((player) => player.memberId === actorMemberId)?.seat ?? -1;

  return {
    roomId: normalizedBoundary.roomId,
    roomCode: normalizedBoundary.roomCode,
    version: normalizedBoundary.version,
    handId: textValue(handSource, ['handId', 'hand_id'], textValue(source, ['handId', 'hand_id'])) || null,
    turnId: textValue(handSource, ['turnId', 'turn_id'], textValue(source, ['turnId', 'turn_id'])) || null,
    phase,
    config: normalizeConfig(first(source, ['config', 'roomConfig', 'room_config'])),
    players: normalizedPlayers,
    buttonSeat: first(handSource, ['buttonSeat', 'button_seat']) === null
      ? null
      : numberValue(handSource, ['buttonSeat', 'button_seat'], numberValue(source, ['buttonSeat', 'button_seat', 'dealerSeat'], -1)) >= 0
        ? numberValue(handSource, ['buttonSeat', 'button_seat'], numberValue(source, ['buttonSeat', 'button_seat', 'dealerSeat']))
        : null,
    actingSeat: explicitActorSeat >= 0
      ? explicitActorSeat
      : derivedActorSeat >= 0
        ? derivedActorSeat
        : numberValue(source, ['actingSeat', 'acting_seat', 'currentSeat'], -1) >= 0
          ? numberValue(source, ['actingSeat', 'acting_seat', 'currentSeat'])
          : null,
    communityCards: normalizeCards(first(handSource, ['board']) ?? first(source, ['communityCards', 'community_cards', 'board'])),
    holeCards: normalizeCards(first(heroPlayerSource, ['holeCards', 'hole_cards']) ?? first(source, ['holeCards', 'hole_cards', 'privateCards', 'private_cards'])),
    pots: (listValue(handSource, ['pots']).length > 0 ? listValue(handSource, ['pots']) : listValue(source, ['pots'])).map((item) => {
      const pot = record(item);
      return {
        amount: numberValue(pot, ['amount', 'chips']),
        eligibleMemberIds: listValue(pot, ['eligibleMemberIds', 'eligible_member_ids'])
          .filter((id): id is string => typeof id === 'string'),
      };
    }),
    currentBet: numberValue(handSource, ['currentBet', 'current_bet'], numberValue(source, ['currentBet', 'current_bet'])),
    legalActions: normalizeLegalActions(first(handSource, ['legalActions', 'legal_actions']) ?? first(source, ['legalActions', 'legal_actions'])),
    deadlineAt: deadline || null,
    serverNow: textValue(source, ['serverNow', 'server_now'], new Date().toISOString()),
    heroMemberId,
    lateJoin: normalizedPlayers.find((player) => player.memberId === heroMemberId)?.status === 'sitting_out'
      || (
        phase === 'playing'
        && normalizedPlayers.find((player) => player.memberId === heroMemberId)?.status === 'waiting'
      )
      || boolValue(source, ['lateJoin', 'late_join', 'waitingForNextHand']),
    paused: phase === 'paused' || boolValue(source, ['paused']),
    closePending: boolValue(source, ['closePending', 'close_pending']),
    settlement: winnerValues.length > 0
      ? {
          winners: winnerValues.map((value) => {
            const winner = record(value);
            return {
              memberId: textValue(winner, ['memberId', 'member_id', 'playerId']),
              nickname: textValue(
                winner,
                ['nickname', 'name'],
                normalizedPlayers.find((player) => player.memberId === textValue(winner, ['memberId', 'member_id', 'playerId']))?.nickname ?? '玩家',
              ),
              amount: numberValue(winner, ['amount', 'won']),
              handName: textValue(winner, ['handName', 'hand_name']) || undefined,
            };
          }),
          reason: textValue(settlementSource, ['reason']) || undefined,
        }
      : undefined,
    roomLogs: normalizeRoomLogs(source),
    completedHands: normalizeCompletedHands(source),
    fairness: {
      serverCommitment: textValue(fairnessSource, ['serverCommitment', 'server_commitment', 'commitment']) || undefined,
      merkleRoot: textValue(fairnessSource, ['merkleRoot', 'merkle_root']) || undefined,
      contributionsReceived: numberValue(fairnessSource, ['contributionsReceived', 'contributions_received']),
      contributionsRequired: numberValue(fairnessSource, ['contributionsRequired', 'contributions_required']),
      contributionRequired: boolValue(fairnessSource, ['contributionRequired', 'contribution_required']),
      auditAvailable: boolValue(fairnessSource, ['auditAvailable', 'audit_available']),
    },
  };
};

export const normalizeRoomSession = (value: unknown, nickname: string): RoomSession => {
  const outer = record(value);
  const source = record(first(outer, ['data', 'room']) ?? outer);
  const session = {
    roomId: textValue(source, ['roomId', 'room_id', 'id']),
    roomCode: textValue(source, ['roomCode', 'room_code', 'code']).toUpperCase(),
    memberId: textValue(source, ['memberId', 'member_id', 'playerId', 'player_id']),
    nickname,
  };
  if (!session.roomId || !session.roomCode || !session.memberId) {
    throw new Error('服务器返回的房间凭据不完整');
  }
  return session;
};

export const normalizeSocketTicket = (value: unknown): SocketTicket => {
  const outer = record(value);
  const source = record(first(outer, ['data']) ?? outer);
  const ticket = textValue(source, ['ticket', 'socketTicket', 'socket_ticket']);
  if (!ticket) throw new Error('服务器未返回实时连接票据');
  return { ticket, expiresAt: textValue(source, ['expiresAt', 'expires_at']) || undefined };
};

export const normalizeAck = (value: unknown, commandId: string): CommandAck => {
  const outer = record(value);
  const source = record(first(outer, ['data', 'ack']) ?? outer);
  const rawStatus = textValue(source, ['status'], boolValue(source, ['accepted']) ? 'accepted' : 'rejected');
  const snapshotValue = first(source, ['snapshot']);
  return {
    status: rawStatus === 'accepted' ? 'accepted' : 'rejected',
    commandId: textValue(source, ['commandId', 'command_id'], commandId),
    appliedVersion: numberValue(source, ['appliedVersion', 'applied_version', 'version']),
    errorCode: textValue(source, ['errorCode', 'error_code']) || undefined,
    message: textValue(source, ['message', 'detail']) || undefined,
    snapshot: snapshotValue ? normalizeSnapshot(snapshotValue) : undefined,
  };
};

export const normalizeRoomEvent = (value: unknown): RoomEvent => {
  const source = record(value);
  const snapshotValue = first(source, ['snapshot']);
  return {
    version: numberValue(source, ['version', 'appliedVersion', 'applied_version']),
    type: textValue(source, ['type', 'eventType', 'event_type'], 'room_changed'),
    snapshot: snapshotValue ? normalizeSnapshot(snapshotValue) : undefined,
  };
};

export const makeCommand = (
  snapshot: GameSnapshot,
  type: RoomCommandType,
  payload: Record<string, unknown> = {},
): RoomCommand => ({
  commandId: crypto.randomUUID(),
  roomId: snapshot.roomId,
  handId: snapshot.handId,
  turnId: snapshot.turnId,
  expectedVersion: snapshot.version,
  type,
  payload,
});

export const DEFAULT_ROOM_CONFIG: RoomConfig = {
  maxPlayers: 8,
  smallBlind: 10,
  bigBlind: 20,
  initialStack: 2000,
  actionTimeoutSeconds: 30,
};

File: apps/server/app/protocol.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, field_validator, model_validator

PROTOCOL_VERSION: Literal["1.0"] = "1.0"
RULES_VERSION: Literal["pokerkit-0.7.5/nlhe-v1"] = "pokerkit-0.7.5/nlhe-v1"


def _to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(word.capitalize() for word in rest)


class ProtocolModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=_to_camel,
        populate_by_name=True,
        serialize_by_alias=True,
        extra="forbid",
    )


class RoomPhase(StrEnum):
    LOBBY = "lobby"
    COLLECTING_ENTROPY = "collecting_entropy"
    PLAYING = "playing"
    SETTLEMENT = "settlement"
    PAUSED = "paused"
    CLOSED = "closed"


class PlayerStatus(StrEnum):
    WAITING = "waiting"
    ACTIVE = "active"
    FOLDED = "folded"
    ALL_IN = "all_in"
    SITTING_OUT = "sitting_out"


class Street(StrEnum):
    PREFLOP = "preflop"
    FLOP = "flop"
    TURN = "turn"
    RIVER = "river"
    COMPLETE = "complete"


class ErrorCode(StrEnum):
    AUTH_REQUIRED = "AUTH_REQUIRED"
    INVALID_COMMAND = "INVALID_COMMAND"
    ROOM_NOT_FOUND = "ROOM_NOT_FOUND"
    ROOM_FULL = "ROOM_FULL"
    ROOM_CLOSED = "ROOM_CLOSED"
    NOT_HOST = "NOT_HOST"
    NOT_MEMBER = "NOT_MEMBER"
    NOT_YOUR_TURN = "NOT_YOUR_TURN"
    INVALID_PHASE = "INVALID_PHASE"
    INVALID_ACTION = "INVALID_ACTION"
    INVALID_AMOUNT = "INVALID_AMOUNT"
    STALE_VERSION = "STALE_VERSION"
    STALE_HAND = "STALE_HAND"
    STALE_TURN = "STALE_TURN"
    ENTROPY_REQUIRED = "ENTROPY_REQUIRED"
    RATE_LIMITED = "RATE_LIMITED"
    DATABASE_UNAVAILABLE = "DATABASE_UNAVAILABLE"


class RoomConfig(ProtocolModel):
    max_players: int = Field(default=8, ge=2, le=10)
    small_blind: int = Field(default=10, ge=1, le=1_000_000)
    big_blind: int = Field(default=20, ge=2, le=2_000_000)
    initial_stack: int = Field(default=2000, ge=1)
    action_timeout_seconds: Literal[15, 30, 60] = 30

    @model_validator(mode="after")
    def validate_blinds_and_stack(self) -> RoomConfig:
        if self.big_blind <= self.small_blind:
            raise ValueError("bigBlind must be greater than smallBlind")
        if not self.big_blind * 20 <= self.initial_stack <= self.big_blind * 500:
            raise ValueError("initialStack must be between 20 and 500 big blinds")
        return self


class CreateRoomRequest(ProtocolModel):
    nickname: str = Field(min_length=1, max_length=20)
    config: RoomConfig = Field(default_factory=RoomConfig)

    @field_validator("nickname")
    @classmethod
    def clean_nickname(cls, value: str) -> str:
        value = " ".join(value.strip().split())
        if not value or any(ord(character) < 32 for character in value):
            raise ValueError("nickname contains unsupported characters")
        return value


class JoinRoomRequest(ProtocolModel):
    nickname: str = Field(min_length=1, max_length=20)

    @field_validator("nickname")
    @classmethod
    def clean_nickname(cls, value: str) -> str:
        return CreateRoomRequest.clean_nickname(value)


class RoomIdentityResponse(ProtocolModel):
    room_id: str
    room_code: str
    member_id: str


class SocketTicketResponse(ProtocolModel):
    ticket: str
    expires_at: datetime


class EmptyPayload(ProtocolModel):
    pass


class ReadyPayload(ProtocolModel):
    ready: bool = True


class PausePayload(ProtocolModel):
    paused: bool = True


class SitOutPayload(ProtocolModel):
    sitting_out: bool = True


class RaiseToPayload(ProtocolModel):
    amount: int = Field(gt=0)


class ShuffleContributionPayload(ProtocolModel):
    entropy: str | None = Field(default=None, min_length=43, max_length=44)
    contribution: str | None = Field(default=None, min_length=64, max_length=64)

    @model_validator(mode="after")
    def require_one_encoding(self) -> ShuffleContributionPayload:
        if (self.entropy is None) == (self.contribution is None):
            raise ValueError("provide exactly one of entropy or contribution")
        return self


class UpdateConfigPayload(ProtocolModel):
    config: RoomConfig


class RemovePlayerPayload(ProtocolModel):
    member_id: str


class PlayerActionPayload(ProtocolModel):
    action: Literal["fold", "check", "call", "raise_to"]
    amount: int | None = Field(default=None, gt=0)


class CommandBase(ProtocolModel):
    command_id: str = Field(min_length=8, max_length=64)
    room_id: str
    hand_id: str | None = None
    turn_id: str | None = None
    expected_version: int = Field(ge=0)


class ReadyCommand(CommandBase):
    type: Literal["ready"]
    payload: ReadyPayload = Field(default_factory=ReadyPayload)


class StartCommand(CommandBase):
    type: Literal["start"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class PauseCommand(CommandBase):
    type: Literal["pause"]
    payload: PausePayload = Field(default_factory=PausePayload)


class CloseCommand(CommandBase):
    type: Literal["close"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class SitOutCommand(CommandBase):
    type: Literal["sitOut"]
    payload: SitOutPayload = Field(default_factory=SitOutPayload)


class FoldCommand(CommandBase):
    type: Literal["fold"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class CheckCommand(CommandBase):
    type: Literal["check"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class CallCommand(CommandBase):
    type: Literal["call"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class RaiseToCommand(CommandBase):
    type: Literal["raiseTo"]
    payload: RaiseToPayload


class ShuffleContributionCommand(CommandBase):
    type: Literal["shuffle.contribute"]
    payload: ShuffleContributionPayload


class SetReadyCommand(CommandBase):
    type: Literal["set_ready"]
    payload: ReadyPayload = Field(default_factory=ReadyPayload)


class StartHandCommand(CommandBase):
    type: Literal["start_hand"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class ContributeRandomnessCommand(CommandBase):
    type: Literal["contribute_randomness"]
    payload: ShuffleContributionPayload


class PlayerActionCommand(CommandBase):
    type: Literal["player_action"]
    payload: PlayerActionPayload


class PauseRoomCommand(CommandBase):
    type: Literal["pause_room"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class ResumeRoomCommand(CommandBase):
    type: Literal["resume_room"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class CloseRoomCommand(CommandBase):
    type: Literal["close_room"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class UpdateConfigCommand(CommandBase):
    type: Literal["update_config"]
    payload: UpdateConfigPayload


class RemovePlayerCommand(CommandBase):
    type: Literal["remove_player"]
    payload: RemovePlayerPayload


class RequestSnapshotCommand(CommandBase):
    type: Literal["request_snapshot"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


class RequestRebuyCommand(CommandBase):
    type: Literal["request_rebuy"]
    payload: EmptyPayload = Field(default_factory=EmptyPayload)


type RoomCommandUnion = (
    ReadyCommand
    | StartCommand
    | PauseCommand
    | CloseCommand
    | SitOutCommand
    | FoldCommand
    | CheckCommand
    | CallCommand
    | RaiseToCommand
    | ShuffleContributionCommand
    | SetReadyCommand
    | StartHandCommand
    | ContributeRandomnessCommand
    | PlayerActionCommand
    | PauseRoomCommand
    | ResumeRoomCommand
    | CloseRoomCommand
    | UpdateConfigCommand
    | RemovePlayerCommand
    | RequestSnapshotCommand
    | RequestRebuyCommand
)
type RoomCommand = Annotated[RoomCommandUnion, Field(discriminator="type")]

ROOM_COMMAND_ADAPTER: TypeAdapter[RoomCommand] = TypeAdapter(RoomCommand)


class CommandAccepted(ProtocolModel):
    status: Literal["accepted"] = "accepted"
    command_id: str
    applied_version: int


class CommandRejected(ProtocolModel):
    status: Literal["rejected"] = "rejected"
    command_id: str
    applied_version: int
    error_code: ErrorCode
    message: str


type CommandAck = CommandAccepted | CommandRejected


class CardView(ProtocolModel):
    code: str
    deck_index: int = Field(ge=0, le=51)
    salt: str
    proof: list[str]


class PlayerView(ProtocolModel):
    member_id: str
    nickname: str
    seat: int = Field(ge=0, le=9)
    stack: int = Field(ge=0)
    borrowed_total: int = Field(default=0, ge=0)
    ranking_score: int
    street_bet: int = Field(ge=0)
    committed: int = Field(ge=0)
    status: PlayerStatus
    ready: bool
    online: bool
    is_host: bool
    rebuy_pending: bool = False
    last_action: str | None = None
    hole_cards: list[CardView] | None = None


class LegalActions(ProtocolModel):
    can_fold: bool = False
    can_check: bool = False
    can_call: bool = False
    can_raise: bool = False
    can_all_in: bool = False
    call_amount: int = Field(default=0, ge=0)
    min_raise_to: int | None = Field(default=None, ge=0)
    max_raise_to: int | None = Field(default=None, ge=0)


class FairnessView(ProtocolModel):
    server_commitment: str | None = None
    merkle_root: str | None = None
    contributions_received: int = Field(default=0, ge=0)
    contributions_required: int = Field(default=0, ge=0)
    contribution_required: bool = False
    audit_available: bool = False


class PotView(ProtocolModel):
    amount: int = Field(ge=0)
    eligible_member_ids: list[str]


class SettlementAward(ProtocolModel):
    member_id: str
    amount: int = Field(gt=0)
    hand_name: str | None = None


class SettlementView(ProtocolModel):
    awards: list[SettlementAward]


class CompletedHandPlayer(ProtocolModel):
    member_id: str
    nickname: str
    seat: int = Field(ge=0, le=9)
    hole_cards: list[CardView]
    hand_name: str
    delta: int
    folded: bool


class CompletedHand(ProtocolModel):
    hand_id: str
    hand_number: int = Field(ge=1)
    completed_at: datetime
    board: list[CardView]
    pot: int = Field(ge=0)
    players: list[CompletedHandPlayer]


class RoomLogEntry(ProtocolModel):
    id: str
    type: str
    message: str
    created_at: datetime
    hand_id: str | None = None


class HandView(ProtocolModel):
    hand_id: str
    hand_number: int = Field(ge=1)
    button_seat: int = Field(ge=0, le=9)
    street: Street
    board: list[CardView]
    pot: int = Field(ge=0)
    pots: list[PotView]
    current_bet: int = Field(ge=0)
    actor_member_id: str | None = None
    actor_seat: int | None = Field(default=None, ge=0, le=9)
    turn_id: str
    deadline_at: datetime | None = None
    legal_actions: LegalActions = Field(default_factory=LegalActions)
    settlement: SettlementView | None = None


class YouView(ProtocolModel):
    member_id: str
    seat: int = Field(ge=0, le=9)
    is_host: bool


class RoomSnapshot(ProtocolModel):
    protocol_version: Literal["1.0"] = PROTOCOL_VERSION
    rules_version: Literal["pokerkit-0.7.5/nlhe-v1"] = RULES_VERSION
    room_id: str
    room_code: str
    version: int = Field(ge=0)
    phase: RoomPhase
    config: RoomConfig
    you: YouView
    players: list[PlayerView]
    hand: HandView | None = None
    room_logs: list[RoomLogEntry] = Field(default_factory=list)
    completed_hands: list[CompletedHand] = Field(default_factory=list)
    fairness: FairnessView = Field(default_factory=FairnessView)
    server_now: datetime
    close_pending: bool = False


class RoomEventMessage(ProtocolModel):
    room_id: str
    version: int = Field(ge=1)
    type: str
    payload: dict[str, Any]
    created_at: datetime


class RoomErrorMessage(ProtocolModel):
    error_code: ErrorCode
    message: str
    command_id: str | None = None


class AuditEvent(ProtocolModel):
    version: int
    type: str
    payload: dict[str, Any]
    created_at: datetime
    previous_hash: str
    hash: str


class AuditHand(ProtocolModel):
    hand_id: str
    hand_number: int
    server_seed: str
    server_commitment: str
    contributions: dict[str, dict[str, Any]]
    deck: list[str]
    leaf_salts: list[str]
    merkle_root: str


class AuditPackage(ProtocolModel):
    schema_version: Literal["1.0"] = PROTOCOL_VERSION
    rules_version: Literal["pokerkit-0.7.5/nlhe-v1"] = RULES_VERSION
    room_id: str
    room_code: str
    closed_at: datetime
    final_event_hash: str
    events: list[AuditEvent]
    hands: list[AuditHand]
    signature_algorithm: Literal["Ed25519"] = "Ed25519"
    signing_public_key: str
    signature: str


class ProtocolContract(ProtocolModel):
    create_room_request: CreateRoomRequest | None = None
    join_room_request: JoinRoomRequest | None = None
    room_identity_response: RoomIdentityResponse | None = None
    socket_ticket_response: SocketTicketResponse | None = None
    room_command: RoomCommandUnion | None = Field(default=None, discriminator="type")
    command_ack: CommandAck | None = None
    room_snapshot: RoomSnapshot | None = None
    room_event: RoomEventMessage | None = None
    room_error: RoomErrorMessage | None = None
    audit_package: AuditPackage | None = None
